### PR TITLE
Switch to gorilla sessions and encrypted cookies

### PR DESCRIPTION
**Pull request by sethvargo**
This PR does a few things:

- Adds a way to process base64 envvars as []byte
- Switches to gorilla/sessions and encrypted cookies
- Switches to gorilla/sessions for realm selection
- Fixes some bugs with realm selection when a user is a member of multiple realms
- Pass flash to the template

Known issues:

- If you change the cookie encryption token, everything breaks and you have to clear cookies to continue. I can create a follow-up issue to track this.
- I didn't modify our flash implementation since, I'll do that in a follow-up since this is already getting large.

### CODE REVIEW

File: cmd/server/main.go [hunk 1]
```diff
@@ -69,6 +70,14 @@ func realMain(ctx context.Context) error {
 		return fmt.Errorf("failed to process config: %w", err)
 	}
 
+	// Setup sessions
+	sessions := sessions.NewCookieStore(config.CookieKeys.AsBytes()...)
```
Comment (sethvargo) [inline cmd/server/main.go:74]: These are the same options as before

File: pkg/config/server_config.go [hunk 2]
```diff
@@ -74,17 +77,6 @@ func NewServerConfig(ctx context.Context) (*ServerConfig, error) {
 	return &config, nil
 }
 
-func (c *ServerConfig) CSRFKey() ([]byte, error) {
```
Comment (sethvargo) [inline pkg/config/server_config.go:77]: Yay - envconfig will do this now

File: pkg/controller/middleware/auth.go [hunk 3]
```diff
@@ -77,84 +82,41 @@ func (h *RequireRealmAdminHandler) Handle(next http.Handler) http.Handler {
...
 
 			flash.FromContext(w, r).Error("Unauthorized")
 			http.Redirect(w, r, "/signout", http.StatusFound)
 			return
 		}
-		next.ServeHTTP(w, r)
-	})
-}
-
-type RequireAdminHandler struct {
```
Comment (sethvargo) [inline pkg/controller/middleware/auth.go:94]: This was replaced with realmadmin

File: pkg/controller/middleware/auth.go [hunk 4]
```diff
@@ -165,22 +127,32 @@ func (h *RequireAuthHandler) Handle(next http.Handler) http.Handler {
...
-			cookie, err := r.Cookie("session")
+			// Get the session
+			session, err := h.sessions.Get(r, "session")
 			if err != nil {
-				return fmt.Errorf("failed to get cookie: %w", err)
+				return fmt.Errorf("failed to get session: %w", err)
+			}
+
+			// Verify firebase cookie value is present
+			if session.Values == nil || session.Values["firebaseCookie"] == nil {
```
Comment (sethvargo) [inline pkg/controller/middleware/auth.go:137]: This is a somewhat annoying side effect of gorilla sessions. Values can be nil and Values is a `map[interface{}][interface{}]`, so the key can also be nil.

File: pkg/controller/realm/index.go [hunk 5]
```diff
@@ -60,6 +78,7 @@ func (c *Controller) HandleIndex() http.Handler {
 
 		// User must select their realm.
 		m := controller.TemplateMapFromContext(ctx)
+		m["flash"] = flash
```
Comment (sethvargo) [inline pkg/controller/realm/index.go:81]: Lost many hours to this.

File: pkg/controller/realm/select.go [hunk 6]
```diff
@@ -23,44 +23,60 @@ import (
...
+		}
+
 		user := controller.UserFromContext(ctx)
 		if user == nil {
 			flash.Error("Unauthorized.")
 			http.Redirect(w, r, "/signout", http.StatusSeeOther)
 			return
 		}
 
-		realm := controller.RealmFromContext(ctx)
```
Comment (sethvargo) [inline pkg/controller/realm/select.go:40]: I'm not entirely sure how this ever worked before? This is the page that sets the realm's context, but it required a context to be set.

File: cmd/server/main.go [hunk 7]
```diff
@@ -69,6 +70,14 @@ func realMain(ctx context.Context) error {
...
+	sessions.Options.Path = "/"
+	sessions.Options.Domain = "" // TODO: make configurable
```
Comment (mikehelmick) [inline cmd/server/main.go:76]: will be interesting to configure in a cloud run + custom domains scenario.

File: cmd/server/main.go [hunk 8]
```diff
@@ -162,31 +175,37 @@ func realMain(ctx context.Context) error {
...
+		sub.Use(requireAuth.Handle)
+		sub.Use(requireRealm.Handle)
+		sub.Use(requireAdmin.Handle)
 
 		apikeyController := apikey.New(ctx, config, db, h)
 		sub.Handle("", apikeyController.HandleIndex()).Methods("GET")
 		sub.Handle("/create", apikeyController.HandleCreate()).Methods("POST")
+	}
 
+	// userrs
```
Comment (mikehelmick) [inline cmd/server/main.go:190]: users

File: pkg/config/server_config.go [hunk 9]
```diff
@@ -41,10 +40,14 @@ type ServerConfig struct {
...
 	RevokeCheckPeriod     time.Duration `env:"REVOKE_CHECK_DURATION,default=5m"`
 
-	// CSRF Secret Key. Must be 32-bytes. Can be generated with tools/gen-secret
-	// Use the syntax of secret:// to pull the secret from secret manager.
-	// We assume the secret itself is base64 encoded. Use CSRFKey() to transform to bytes.
-	CSRFAuthKey string `env:"CSRF_AUTH_KEY,required"`
+	// CookieKeys is a slice of bytes. The odd values are hash keys to HMAC the
+	// cookies. The even values are block keys to encrypt the cookie. Both keys
+	// should be 64 bytes. The value's should be specified as base64 encoded.
+	CookieKeys Base64ByteSlice `env:"COOKIE_KEYS,required"`
```
Comment (mikehelmick) [inline pkg/config/server_config.go:46]: nice 👍

File: pkg/controller/middleware/auth.go (hunk 3 above)
Comment (mikehelmick) [reply pkg/controller/middleware/auth.go:94]: We still need the require admin handler for the ability to create/delete realms - can I haz back?

File: pkg/controller/realm/select.go (hunk 6 above)
Comment (mikehelmick) [reply pkg/controller/realm/select.go:40]: RealmFromContext returned nil if one wasn't set - so this worked fine

Comment (mikehelmick) [reply pkg/controller/realm/select.go:40]: ah - no this got changed

I had is reading the cookie directly.

File: cmd/server/main.go (hunk 7 above)
Comment (sethvargo) [reply cmd/server/main.go:76]: It has to match what the user hits in the browser.

**Comment by sethvargo** [comment]
Fixed typo and rebased

File: pkg/controller/middleware/auth.go (hunk 3 above)
Comment (sethvargo) [reply pkg/controller/middleware/auth.go:94]: Since you asked nicely...

File: pkg/config/server_config.go (hunk 9 above)
Comment (sethvargo) [reply pkg/config/server_config.go:46]: We have to pull in the changes after https://github.com/google/exposure-notifications-server/pull/788#event-3608455621 for it to work with secrets thought. It'll work with local keys just fine.

Comment (sethvargo) [reply pkg/config/server_config.go:46]: https://github.com/google/exposure-notifications-verification-server/pull/159

File: cmd/server/main.go (hunk 7 above)
Comment (sethvargo) [reply cmd/server/main.go]: Made configurable

**Pull request merged**